### PR TITLE
refactor: native zlib support

### PR DESCRIPTION
**Pull request by didinele**
**Please describe the changes this PR makes and why it should be merged:**
#9279 again, but supports both

### CODE REVIEW

File: packages/ws/src/ws/WebSocketShard.ts [hunk 1]
```diff
@@ -543,41 +580,61 @@ export class WebSocketShard extends AsyncEventEmitter<WebSocketShardEventsMap> {
...
+				) as GatewayReceivePayload;
+			}
 		}
 
 		this.debug([
 			'Received a message we were unable to decompress',
 			`isBinary: ${isBinary.toString()}`,
-			`useIdentifyCompress: ${this.useIdentifyCompress.toString()}`,
-			`inflate: ${Boolean(this.inflate).toString()}`,
+			`useIdentifyCompress: ${this.useIdentifyCompression.toString()}`,
```
Comment (SuperchupuDev) [inline packages/ws/src/ws/WebSocketShard.ts:636]:
```suggestion
			`useIdentifyCompression: ${this.useIdentifyCompression.toString()}`,
```
same thing i suggested in the previous pr :P

File: packages/ws/src/ws/WebSocketShard.ts [hunk 2]
```diff
@@ -543,41 +580,61 @@ export class WebSocketShard extends AsyncEventEmitter<WebSocketShardEventsMap> {
...
+			`inflate: ${Boolean(this.nativeInflate).toString()}`,
```
[the remark this answers is not in the document]
Comment (didinele) [reply packages/ws/src/ws/WebSocketShard.ts:637]: oh, thanks. will do this a little different though

File: packages/ws/README.md [hunk 3]
```diff
@@ -50,7 +50,10 @@ const manager = new WebSocketManager({
 	intents: 0, // for no intents
 	rest,
 	// uncomment if you have zlib-sync installed and want to use compression
-	// compression: CompressionMethod.ZlibStream,
+	// compression: CompressionMethod.ZLibSync,
+
+	// alternatively, we now support compression using node's native `node:zlib` module:
```
Comment (Jiralite) [inline packages/ws/README.md:55]: Words like "now" don't bode well in the future.

```suggestion
	// alternatively, we support compression using node's native `node:zlib` module:
```

File: packages/ws/src/utils/constants.ts [hunk 4]
```diff
@@ -18,13 +18,19 @@ export enum Encoding {
...
 }
 
 export const DefaultDeviceProperty = `@discordjs/ws [VI]{{inject}}[/VI]` as `@discordjs/ws ${string}`;
 
 const getDefaultSessionStore = lazy(() => new Collection<number, SessionInfo | null>());
 
+export const CompressionParameterMap = {
+	[CompressionMethod.ZLibNative]: 'zlib-stream',
+	[CompressionMethod.ZLibSync]: 'zlib-stream',
+} as const satisfies Record<CompressionMethod, string>;
```
Comment (Jiralite) [inline packages/ws/src/utils/constants.ts:32]: Isn't this a `Readonly<Record<...>>`?

Comment (almeidx) [reply packages/ws/src/utils/constants.ts:32]: `satisfies` does not alter the type of the variable, it is merely used for validating that the object conforms to that type. Besides, `as const` already makes the type readonly.

Comment (Jiralite) [reply packages/ws/src/utils/constants.ts:32]: What I meant was wouldn't it be more ideal to satisfy something that's `as const` (which is immutable) with something that is not immutable? So it should be like:

```suggestion
} as const satisfies Readonly<Record<CompressionMethod, string>>;
```

Comment (didinele) [reply packages/ws/src/utils/constants.ts:32]: it's not any better IMO, just more verbose for no good reason

File: packages/ws/src/utils/constants.ts [hunk 5]
```diff
@@ -18,13 +18,19 @@ export enum Encoding {
  * Valid compression methods
  */
 export enum CompressionMethod {
-	ZlibStream = 'zlib-stream',
+	ZLibNative,
```
Comment (vladfrangu) [inline packages/ws/src/utils/constants.ts:21]: Why the capitalized `L` 😢

File: packages/ws/src/utils/constants.ts (hunk 4 above)
Comment (vladfrangu) [reply packages/ws/src/utils/constants.ts:32]: +1, the exported type is still going to be marked as readonly

File: packages/ws/src/ws/WebSocketShard.ts [hunk 6]
```diff
@@ -451,28 +485,29 @@ export class WebSocketShard extends AsyncEventEmitter<WebSocketShardEventsMap> {
 			`shard id: ${this.id.toString()}`,
 			`shard count: ${this.strategy.options.shardCount}`,
 			`intents: ${this.strategy.options.intents}`,
-			`compression: ${this.inflate ? 'zlib-stream' : this.useIdentifyCompress ? 'identify' : 'none'}`,
+			`compression: ${this.strategy.options.compression === null ? (this.useIdentifyCompression ? 'identify' : 'none') : 'zlib-stream'}`,
```
Comment (vladfrangu) [inline packages/ws/src/ws/WebSocketShard.ts:488]: This should use the compression map, not a hardcoded string for the non-useIdentifyCompression path

File: packages/ws/src/ws/WebSocketShard.ts [hunk 7]
```diff
@@ -543,41 +580,61 @@ export class WebSocketShard extends AsyncEventEmitter<WebSocketShardEventsMap> {
...
+				decompressable.length >= 4 &&
+				decompressable.at(-4) === 0x00 &&
+				decompressable.at(-3) === 0x00 &&
+				decompressable.at(-2) === 0xff &&
+				decompressable.at(-1) === 0xff;
 
-			const zlib = (await getZlibSync())!;
-			this.inflate.push(Buffer.from(decompressable), flush ? zlib.Z_SYNC_FLUSH : zlib.Z_NO_FLUSH);
+			if (this.nativeInflate) {
+				this.nativeInflate!.write(decompressable, 'binary');
```
Comment (vladfrangu) [inline packages/ws/src/ws/WebSocketShard.ts:592]: And if that `!` is needed, I will yell at TS

```suggestion
				this.nativeInflate.write(decompressable, 'binary');
```

File: packages/ws/src/ws/WebSocketShard.ts [hunk 8]
```diff
@@ -543,41 +580,61 @@ export class WebSocketShard extends AsyncEventEmitter<WebSocketShardEventsMap> {
...
+			}
 		}
 
 		this.debug([
 			'Received a message we were unable to decompress',
 			`isBinary: ${isBinary.toString()}`,
-			`useIdentifyCompress: ${this.useIdentifyCompress.toString()}`,
-			`inflate: ${Boolean(this.inflate).toString()}`,
+			`useIdentifyCompression: ${this.useIdentifyCompression.toString()}`,
+			`inflate: ${this.nativeInflate ? 'node:zlib' : 'zlib-sync'} (zlib-stream)`,
```
Comment (vladfrangu) [inline packages/ws/src/ws/WebSocketShard.ts:637]: Should make this a map too ideally, so if/when we add zstd-stream support in we dont need a ton of extra changes

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: packages/ws/src/utils/constants.ts (hunk 5 above)
Comment (didinele) [reply packages/ws/src/utils/constants.ts:21]: dunno. felt right to me. if you want, we can change it back

File: packages/ws/src/ws/WebSocketShard.ts (hunk 8 above)
Comment (didinele) [reply packages/ws/src/ws/WebSocketShard.ts:637]: I have a different idea for this, let me know if you think this is ok.

File: packages/ws/src/utils/constants.ts (hunk 5 above)
Comment (didinele) [reply packages/ws/src/utils/constants.ts:21]: We settled on `Zlib`, vote was here:

- https://canary.discord.com/channels/222078108977594368/1209960895141388389/1236249988871229461

File: packages/ws/src/ws/WebSocketShard.ts [hunk 9]
```diff
@@ -161,21 +179,63 @@ export class WebSocketShard extends AsyncEventEmitter<WebSocketShardEventsMap> {
...
+						console.warn('WebSocketShard: Compression is set to zlib-sync, but it is not installed.');
+						params.delete('compress');
+					}
+
+					break;
+				}
+			}
+		}
+
+		if (useIdentifyCompression) {
```
[the remark this answers is not in the document]
Comment (didinele) [reply packages/ws/src/ws/WebSocketShard.ts:234]: good catch

File: packages/ws/src/ws/WebSocketShard.ts [hunk 10]
```diff
@@ -120,6 +126,18 @@ export class WebSocketShard extends AsyncEventEmitter<WebSocketShardEventsMap> {
...
+	private identifyCompressionEnabled = false;
+
+	/**
+	 * @privateRemarks
+	 *
+	 * This is needed because `this.strategy.options.compression` is not an actual reflection of the compression method
+	 * used, but rather the compression method that the user wants to use. This is because the libraries could just be missing.
+	 */
+	private get transportCompressionEnabled() {
+		return this.strategy.options.compression !== null && (Boolean(this.nativeInflate) || Boolean(this.zLibSyncInflate));
```
Comment (kyranet) [inline packages/ws/src/ws/WebSocketShard.ts:138]: Single `Boolean` call:
```suggestion
		return this.strategy.options.compression !== null && Boolean(this.nativeInflate || this.zLibSyncInflate);
```

Alternatively:
```suggestion
		return this.strategy.options.compression !== null && (this.nativeInflate ?? this.zLibSyncInflate) !== null;
```

**Review by kyranet — COMMENTED** [review]
Just one small thing, not blocking

**Review by kyranet — APPROVED** [review]

**Review by vladfrangu — APPROVED** [review]

File: packages/ws/src/ws/WebSocketShard.ts [hunk 11]
```diff
@@ -542,42 +615,50 @@ export class WebSocketShard extends AsyncEventEmitter<WebSocketShardEventsMap> {
...
+				return this.parseInflateResult(result);
+			}
 		}
 
 		this.debug([
 			'Received a message we were unable to decompress',
 			`isBinary: ${isBinary.toString()}`,
-			`useIdentifyCompress: ${this.useIdentifyCompress.toString()}`,
-			`inflate: ${Boolean(this.inflate).toString()}`,
+			`useIdentifyCompression: ${this.identifyCompressionEnabled.toString()}`,
```
Comment (SuperchupuDev) [inline packages/ws/src/ws/WebSocketShard.ts:660]:
```suggestion
			`identifyCompressionEnabled: ${this.identifyCompressionEnabled.toString()}`,
```
missed it again :P

**Review by SpaceEEC — APPROVED** [review]

**Pull request merged**